src/libp2p_transport_p2p.erl

```diff
@@ -63,11 +64,17 @@ connect_to(MAddr, UserOptions, Timeout, TID) ->
                                 {ok, SessionPid}->
                                     libp2p_config:insert_session(TID, MAddr, SessionPid),
                                     {ok, SessionPid};
-                                {error, Error} -> {error, Error}
+                                {error, Error} ->
+                                    libp2p_peerbook:refresh(Peerbook, PeerInfo),
+                                    {error, Error}
                             end;
-                        {error, Error} -> {error, Error}
+                        {error, Error} ->
+                            libp2p_peerbook:refresh(Peerbook, PeerInfo),
+                            {error, Error}
                     end;
-                {error, Reason} -> {error, Reason}
+                {error, Reason} ->
+                    libp2p_peerbook:refresh(Peerbook, Addr),
+                    {error, Reason}
```

Comment:
Instead of repeating the refresh many times, maybe it would be better to handle it like this:

```erlang
  Response = case libp2p_peerbook:get(Peerbook, Addr) of
    %% ...
  end,
  case Response of
    {error, _Error} ->
      libp2p_peerbook:refresh(Peerbook, PeerInfo);
    _OK -> ok
  end,
  Response;
```